File: programmers/level1/67256.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

string solution(vector<int> numbers, string hand) {
    string answer = "";

    pair<int, int> l = {3, 0};
    pair<int, int> r = {3, 2};

    for (int i = 0; i < numbers.size(); i++) {
        int num = numbers[i];
        if (num == 1 || num == 4 || num == 7) {
            answer += 'L';
            if (num == 1) {
                l = {0, 0};
            } else if (num == 4) {
                l = {1, 0};
            } else if (num == 7) {
                l = {2, 0};
            }
        } else if (num == 3 || num == 6 || num == 9) {
            answer += 'R';
            if (num == 3) {
                r = {0, 2};
            } else if (num == 6) {
                r = {1, 2};
            } else if (num == 9) {
                r = {2, 2};
            }
        } else if (num == 2 || num == 5 || num == 8 || num == 0) {
            if(num == 2) {
                pair<int, int> cur = {0, 1};
                int ldiff = abs(cur.first - l.first) + abs(cur.second - l.second);
                int rdiff = abs(cur.first - r.first) + abs(cur.second - r.second);
                if(ldiff < rdiff) {
                    answer += 'L';
                    l = cur;
                } else if(rdiff < ldiff) {
                    answer += 'R';
                    r = cur;
                } else {
                    if(hand == "right") {
                        answer += 'R';
                        r = cur;
                    } else {
                        answer += 'L';
                        l = cur;
                    }
                }
            } else if(num == 5) {
                pair<int, int> cur = {1, 1};
                int ldiff = abs(cur.first - l.first) + abs(cur.second - l.second);
                int rdiff = abs(cur.first - r.first) + abs(cur.second - r.second);
                if(ldiff < rdiff) {
                    answer += 'L';
                    l = cur;
                } else if(rdiff < ldiff) {
                    answer += 'R';
                    r = cur;
                } else {
                    if(hand == "right") {
                        answer += 'R';
                        r = cur;
                    } else {
                        answer += 'L';
                        l = cur;
                    }
                }
            } else if(num == 8) {
                pair<int, int> cur = {2, 1};
                int ldiff = abs(cur.first - l.first) + abs(cur.second - l.second);
                int rdiff = abs(cur.first - r.first) + abs(cur.second - r.second);
                if(ldiff < rdiff) {
                    answer += 'L';
                    l = cur;
                } else if(rdiff < ldiff) {
                    answer += 'R';
                    r = cur;
                } else {
                    if(hand == "right") {
                        answer += 'R';
                        r = cur;
                    } else {
                        answer += 'L';
                        l = cur;
                    }
                }
            } else if(num == 0) {
                pair<int, int> cur = {3, 1};
                int ldiff = abs(cur.first - l.first) + abs(cur.second - l.second);
                int rdiff = abs(cur.first - r.first) + abs(cur.second - r.second);
                if(ldiff < rdiff) {
                    answer += 'L';
                    l = cur;
                } else if(rdiff < ldiff) {
                    answer += 'R';
                    r = cur;
                } else {
                    if (hand == "right") {
                        answer += 'R';
                        r = cur;
                    } else {
                        answer += 'L';
                        l = cur;
                    }
                }
            }
        }
    }

    return answer;
}

File: kotlin_ps/src/main/kotlin/programmers/programmers_67256/67256.kt
package programmers.programmers_67256

import kotlin.math.*

class Solution {
    fun solution(numbers: IntArray, hand: String): String {
        var answer = ""

        var leftHandPosition = Pair(3, 0)
        var rightHandPosition = Pair(3, 2)

        for (number in numbers) {
            val numberPosition = Position.positionOf(number)

            if (numberPosition.second == 0) {
                answer += "L"
                leftHandPosition = numberPosition
            } else if (numberPosition.second == 2) {
                answer += "R"
                rightHandPosition = numberPosition
            } else {
                val leftDistance = getDistance(numberPosition, leftHandPosition)
                val rightDistance = getDistance(numberPosition, rightHandPosition)

                if (leftDistance < rightDistance) {
                    answer += "L"
                    leftHandPosition = numberPosition
                } else if (leftDistance > rightDistance) {
                    answer += "R"
                    rightHandPosition = numberPosition
                } else if (leftDistance == rightDistance) {
                    if (hand == "left") {
                        answer += "L"
                        leftHandPosition = numberPosition
                    } else if (hand == "right") {
                        answer += "R"
                        rightHandPosition = numberPosition
                    }
                }
            }
        }

        return answer
    }

    private fun getDistance(from: Pair<Int, Int>, to: Pair<Int, Int>): Int =
        max(from.first, to.first) - min(from.first, to.first) +
                max(from.second, to.second) - min(from.second, to.second)

    enum class Position(val number: Int, val position: Pair<Int, Int>) {
        ONE(1, Pair(0, 0)),
        TWO(2, Pair(0, 1)),
        THREE(3, Pair(0, 2)),
        FOUR(4, Pair(1, 0)),
        FIVE(5, Pair(1, 1)),
        SIX(6, Pair(1, 2)),
        SEVEN(7, Pair(2, 0)),
        EIGHT(8, Pair(2, 1)),
        NINE(9, Pair(2, 2)),
        ZERO(0, Pair(3, 1));

        companion object {
            fun positionOf(number: Int): Pair<Int, Int> {
                return values().find { it.number == number }!!.position
            }
        }
    }
}